hw/ip/rom_ctrl/util/scramble_image.py

```diff
@@ -458,13 +485,34 @@ def add_hash(self, scr_mem: MemFile) -> None:
         digest_bytes = hash_obj.read(bytes_per_word * num_digest_words)
         digest256 = int.from_bytes(digest_bytes, byteorder='little')
 
-        # Insert the hash back into scr_mem
+        # Chop the 256-bit digest into 32-bit words. These words should never
+        # be read "unscrambled": the rom_ctrl checker reads them raw. We can
+        # guarantee this by fiddling around with the top 7 bits (which are
+        # otherwise ignored) to ensure that they unscramble to words with
+        # invalid ECC checksums.
+        mask32 = (1 << 32) - 1
+        first_digest_idx = self.rom_size_words - num_digest_words
         for digest_idx in range(num_digest_words):
-            log_addr = self.rom_size_words - num_digest_words + digest_idx
-            phy_addr = subst_perm_enc(log_addr, addr_scr_nonce,
-                                      self._addr_width, subst_perm_rounds)
-            digest_word = (digest256 >> (32 * digest_idx)) & ((1 << 32) - 1)
-            scr_chunk.words[phy_addr] = digest_word
+            log_addr = first_digest_idx + digest_idx
+            w32 = (digest256 >> (32 * digest_idx)) & mask32
+            found_mismatch = False
+
+            for chk_bits in range(128):
+                w39 = w32 | (chk_bits << 32)
+                clr39 = self.unscramble_word(39, log_addr, w39)
+                clr32 = clr39 & mask32
+                exp39 = ecc_encode_some('hsiao', 32, [clr32])[0][0]
```

Comment:
you may want to change this to `inv_hsiao` now.

Reply:
Ahah, thanks! I was wondering which of us would get merged first :-)